File: v1.0/binding/nodejs/ip2region.js
/**
 * ip2region client for nodejs
 *
 * project: https://github.com/lionsoul2014/ip2region
 *
 * @author dongyado<dongyado@gmail.com>
 * @author leeching<leeching.fx@gmail.com>
 * @author dongwei<maledong_github@outlook.com>
 */
const fs = require('fs');
const path = require('path');
const DEFAULT_DB_PATH  = path.join(__dirname,'./data/ip2region.db') ;

//#region Private Functions
/**
 * Convert ip to long (xxx.xxx.xxx.xxx to a integer)
 *
 * @param {string} ip
 * @return {number} long value
 */
function _ip2long(ip) {
    const arr = ip.split('.');
    if (arr.length !== 4) {
        throw new Error('invalid ip');
    }
    return arr.reduce((val, n, i) => {
        n = Number(n);
        if (!Number.isInteger(n) || n < 0 || n > 255) {
            throw new Error('invalid ip');
        }
        return val + IP_BASE[i] * n;
    }, 0);
}

/**
 * Get long value from buffer with specified offset
 *
 * @param {Buffer} buffer
 * @param {number} offset
 * @return {number} long value
 */
function _getLong(buffer, offset) {
    const val =
        (buffer[offset] & 0x000000ff) |
        ((buffer[offset + 1] << 8) & 0x0000ff00) |
        ((buffer[offset + 2] << 16) & 0x00ff0000) |
        ((buffer[offset + 3] << 24) & 0xff000000);
    return val < 0 ? val >>> 0 : val;
}

//#endregion

//#region Private Variables

// We don't wanna expose a private global settings to
// the public for safety reason.
const _globalInstances = new Map();

const IP_BASE = [16777216, 65536, 256, 1];
const INDEX_BLOCK_LENGTH = 12;
const TOTAL_HEADER_LENGTH = 8192;

// Private Message Symbols for functions
const PrepareHeader = Symbol('#PrepareHeader');
const CalTotalBlocks = Symbol('#CalsTotalBlocks');
const ReadDataSync = Symbol('#ReadDataSync');
const ReadData = Symbol('#ReadData');

//#endregion
class IP2Region {

    //#region Private Functions

    [CalTotalBlocks]() {
        const superBlock = Buffer.alloc(8);
        fs.readSync(this.dbFd, superBlock, 0, 8, 0);
        this.firstIndexPtr = _getLong(superBlock, 0);
        this.lastIndexPtr = _getLong(superBlock, 4);
        this.totalBlocks =
            (this.lastIndexPtr - this.firstIndexPtr) / INDEX_BLOCK_LENGTH + 1;
    }

    [PrepareHeader]() {
        fs.readSync(
            this.dbFd,
            this.headerIndexBuffer,
            0,
            TOTAL_HEADER_LENGTH,
            8
        );

        for (let i = 0; i < TOTAL_HEADER_LENGTH; i += 8) {
            const startIp = _getLong(this.headerIndexBuffer, i);
            const dataPtr = _getLong(this.headerIndexBuffer, i + 4);
            if (dataPtr == 0) break;

            this.headerSip.push(startIp);
            this.headerPtr.push(dataPtr);
            this.headerLen++; // header index size count
        }
    }

    [ReadData](dataPos, callBack) {
        if (dataPos == 0) return callBack(null, null);
        const dataLen = (dataPos >> 24) & 0xff;
        dataPos = dataPos & 0x00ffffff;
        const dataBuffer = Buffer.alloc(dataLen);

        fs.read(this.dbFd, dataBuffer, 0, dataLen, dataPos, (err, result) => {
            if (err) {
                callBack(err, null);
            }
            else {
                const city = _getLong(dataBuffer, 0);
                const region = dataBuffer.toString('utf8', 4, dataLen);
                callBack(null, { city, region });
            }
        });
    }

    [ReadDataSync](dataPos) {
        if (dataPos == 0) return null;
        const dataLen = (dataPos >> 24) & 0xff;
        dataPos = dataPos & 0x00ffffff;
        const dataBuffer = Buffer.alloc(dataLen);

        fs.readSync(this.dbFd, dataBuffer, 0, dataLen, dataPos);

        const city = _getLong(dataBuffer, 0);
        const region = dataBuffer.toString('utf8', 4, dataLen);

        return { city, region };
    }
    //#endregion

    //#region Static Functions

    // Single Instance
    static create(dbPath = DEFAULT_DB_PATH) {
        let existInstance = _globalInstances.get(dbPath);
        if (existInstance == null) {
            existInstance = new IP2Region({ dbPath: dbPath });
        }
        return existInstance;
    }

    /**
     * For backward compatibility
     */
    static destroy() {
        _globalInstances.forEach(([key, instance]) => {
            instance.destroy();
        });
    }

    //#endregion

    constructor(options = {}) {

        const { dbPath } = options;

        // Keep for MemorySearch
        this.totalInMemoryBytesSize = fs.statSync(dbPath).size;
        this.totalInMemoryBytes = null;

        this.dbFd = fs.openSync(dbPath, 'r');

        this.dbPath = dbPath;

        _globalInstances.set(this.dbPath, this);

        this.totalBlocks = this.firstIndexPtr = this.lastIndexPtr = 0;
        this[CalTotalBlocks]();

        this.headerIndexBuffer = Buffer.alloc(TOTAL_HEADER_LENGTH);
        this.headerSip = [];
        this.headerPtr = [];
        this.headerLen = 0;
        this[PrepareHeader]();
    }

    //#region Public Functions
    /**
     * Destroy the current file by closing it.
     */
    destroy() {
        fs.closeSync(this.dbFd);
        _globalInstances.delete(this.dbPath);
    }

    /**
     * Sync of binarySearch.
     * @param {string} ip The IP address to search for.
     * @return {SearchResult} A result something like `{ city: 2163, region: '中国|0|广东省|深圳市|阿里云' }`
     */
    binarySearchSync(ip) {

        ip = _ip2long(ip);

        let low = 0;
        let mid = 0;
        let high = this.totalBlocks;
        let pos = 0;
        let sip = 0;
        const indexBuffer = Buffer.alloc(12);

        // binary search
        while (low <= high) {
            mid = (low + high) >> 1;
            pos = this.firstIndexPtr + mid * INDEX_BLOCK_LENGTH;
            fs.readSync(this.dbFd, indexBuffer, 0, INDEX_BLOCK_LENGTH, pos);
            sip = _getLong(indexBuffer, 0);

            if (ip < sip) {
                high = mid - 1;
            } else {
                sip = _getLong(indexBuffer, 4);
                if (ip > sip) {
                    low = mid + 1;
                } else {
                    sip = _getLong(indexBuffer, 8);
                    break;
                }
            }
        }
        return this[ReadDataSync](sip);
    }

    /**
     * Async of binarySearch.
     * @param {string} ip The IP address to search for.
     * @param {Function} callBack The callBack function with two parameters, if successful,
     * err is null and result is `{ city: 2163, region: '中国|0|广东省|深圳市|阿里云' }`
     */
    binarySearch(ip, callBack) {

        ip = _ip2long(ip);

        let low = 0;
        let mid = 0;
        let high = this.totalBlocks;
        let pos = 0;
        let sip = 0;
        const indexBuffer = Buffer.alloc(12);
        const _self = this;

        // Because `while` is a sync method, we have to convert this to a recursive loop
        // and in each loop we should continue calling `setImmediate` until we found the IP.
        function _innerAsyncWhile() {
            if (low <= high) {
                mid = (low + high) >> 1;
                pos = _self.firstIndexPtr + mid * INDEX_BLOCK_LENGTH;

                // Now async read the file
                fs.read(_self.dbFd, indexBuffer, 0, INDEX_BLOCK_LENGTH, pos, (err) => {

                    if (err) {
                        return callBack(err, null);
                    }

                    sip = _getLong(indexBuffer, 0);

                    if (ip < sip) {
                        high = mid - 1;
                        setImmediate(_innerAsyncWhile);
                    } else {
                        sip = _getLong(indexBuffer, 4);
                        if (ip > sip) {
                            low = mid + 1;
                            setImmediate(_innerAsyncWhile);
                        } else {
                            sip = _getLong(indexBuffer, 8);
                            _self[ReadData](sip, (err, result) => {
                                callBack(err, result);
                            });
                        }
                    }
                });
            }
        }

        // Call this immediately
        _innerAsyncWhile();

    }

    /**
     * Sync of btreeSearch.
     * @param {string} ip The IP address to search for.
     * @return {Function} A result something like `{ city: 2163, region: '中国|0|广东省|深圳市|阿里云' }`
     */
    btreeSearchSync(ip) {

        ip = _ip2long(ip);

        // first search  (in header index)
        let low = 0;
        let mid = 0;
        let high = this.headerLen;
        let sptr = 0;
        let eptr = 0;

        while (low <= high) {
            mid = (low + high) >> 1;

            if (ip == this.headerSip[mid]) {
                if (mid > 0) {
                    sptr = this.headerPtr[mid - 1];
                    eptr = this.headerPtr[mid];
                } else {
                    sptr = this.headerPtr[mid];
                    eptr = this.headerPtr[mid + 1];
                }
                break;
            }

            if (ip < this.headerSip[mid]) {
                if (mid == 0) {
                    sptr = this.headerPtr[mid];
                    eptr = this.headerPtr[mid + 1];
                    break;
                } else if (ip > this.headerSip[mid - 1]) {
                    sptr = this.headerPtr[mid - 1];
                    eptr = this.headerPtr[mid];
                    break;
                }
                high = mid - 1;
            } else {
                if (mid == this.headerLen - 1) {
                    sptr = this.headerPtr[mid - 1];
                    eptr = this.headerPtr[mid];
                    break;
                } else if (ip <= this.headerSip[mid + 1]) {
                    sptr = this.headerPtr[mid];
                    eptr = this.headerPtr[mid + 1];
                    break;
                }
                low = mid + 1;
            }
        }

        // match nothing
        if (sptr == 0) return null;

        // second search (in index)
        const blockLen = eptr - sptr;
        const blockBuffer = Buffer.alloc(blockLen + INDEX_BLOCK_LENGTH);
        fs.readSync(
            this.dbFd,
            blockBuffer,
            0,
            blockLen + INDEX_BLOCK_LENGTH,
            sptr
        );

        low = 0;
        high = blockLen / INDEX_BLOCK_LENGTH;

        let p = 0;
        let sip = 0;

        while (low <= high) {
            mid = (low + high) >> 1;
            p = mid * INDEX_BLOCK_LENGTH;
            sip = _getLong(blockBuffer, p);

            if (ip < sip) {
                high = mid - 1;
            } else {
                sip = _getLong(blockBuffer, p + 4);
                if (ip > sip) {
                    low = mid + 1;
                } else {
                    sip = _getLong(blockBuffer, p + 8);
                    break;
                }
            }
        }
        return this[ReadDataSync](sip);
    }

    /**
    * Async of btreeSearch.
     * @param {string} ip The IP address to search for.
     * @param {Function} callBack The callBack function with two parameters, if successful,
     * err is null and result is `{ city: 2163, region: '中国|0|广东省|深圳市|阿里云' }`
    */
    btreeSearch(ip, callBack) {
        ip = _ip2long(ip);

        // first search  (in header index)
        let low = 0;
        let mid = 0;
        let high = this.headerLen;
        let sptr = 0;
        let eptr = 0;

        while (low <= high) {
            mid = (low + high) >> 1;

            if (ip == this.headerSip[mid]) {
                if (mid > 0) {
                    sptr = this.headerPtr[mid - 1];
                    eptr = this.headerPtr[mid];
                } else {
                    sptr = this.headerPtr[mid];
                    eptr = this.headerPtr[mid + 1];
                }
                break;
            }

            if (ip < this.headerSip[mid]) {
                if (mid == 0) {
                    sptr = this.headerPtr[mid];
                    eptr = this.headerPtr[mid + 1];
                    break;
                } else if (ip > this.headerSip[mid - 1]) {
                    sptr = this.headerPtr[mid - 1];
                    eptr = this.headerPtr[mid];
                    break;
                }
                high = mid - 1;
            } else {
                if (mid == this.headerLen - 1) {
                    sptr = this.headerPtr[mid - 1];
                    eptr = this.headerPtr[mid];
                    break;
                } else if (ip <= this.headerSip[mid + 1]) {
                    sptr = this.headerPtr[mid];
                    eptr = this.headerPtr[mid + 1];
                    break;
                }
                low = mid + 1;
            }
        }

        // match nothing
        if (sptr == 0) return callBack(null, null);

        let p = 0;
        let sip = 0;

        // second search (in index)
        const blockLen = eptr - sptr;
        const blockBuffer = Buffer.alloc(blockLen + INDEX_BLOCK_LENGTH);
        low = 0;
        high = blockLen / INDEX_BLOCK_LENGTH;

        const _self = this;

        function _innerAsyncWhile() {

            if (low <= high) {

                mid = (low + high) >> 1;
                p = mid * INDEX_BLOCK_LENGTH;

                // Use this to call the method itself as
                // an asynchronize step
                fs.read(_self.dbFd, blockBuffer,
                    0,
                    blockLen + INDEX_BLOCK_LENGTH,
                    sptr, (err) => {

                        if (err) {
                            return callBack(err, null);
                        }

                        sip = _getLong(blockBuffer, p);

                        if (ip < sip) {
                            high = mid - 1;
                            setImmediate(_innerAsyncWhile);
                        } else {
                            sip = _getLong(blockBuffer, p + 4);
                            if (ip > sip) {
                                low = mid + 1;
                                setImmediate(_innerAsyncWhile);
                            } else {
                                sip = _getLong(blockBuffer, p + 8);
                                _self[ReadData](sip, (err, result) => {
                                    callBack(err, result);
                                });
                            }
                        }
                    });
            }
            else {
                // If we found nothing, return null
                return callBack(null, null);
            }
        }

        _innerAsyncWhile();
    }

    /**
     * Sync of MemorySearch.
     * @param {String} ip 
     */
    memorySearchSync(ip) {

        ip = _ip2long(ip);

        if (this.totalInMemoryBytes === null) {

            this.totalInMemoryBytes = Buffer.alloc(this.totalInMemoryBytesSize);
            fs.readSync(this.dbFd, this.totalInMemoryBytes, 0, this.totalInMemoryBytesSize, 0);

            this.firstIndexPtr = _getLong(this.totalInMemoryBytes, 0);
            this.lastIndexPtr = _getLong(this.totalInMemoryBytes, 4);
            this.totalBlocks = ((this.lastIndexPtr - this.firstIndexPtr) / INDEX_BLOCK_LENGTH) | 0 + 1;
        }

        let l = 0, h = this.totalBlocks;
        let sip = 0;
        let m = 0, p = 0;

        while (l <= h) {
            m = (l + h) >> 1;
            p = (this.firstIndexPtr + m * INDEX_BLOCK_LENGTH) | 0;

            sip = _getLong(this.totalInMemoryBytes, p);

            if (ip < sip) {
                h = m - 1;
            }
            else {
                sip = _getLong(this.totalInMemoryBytes, p + 4);
                if (ip > sip) {
                    l = m + 1;
                }
                else {
                    sip = _getLong(this.totalInMemoryBytes, p + 8);
                    //not matched
                    if (sip === 0) return null;

                    //get the data
                    let dataLen = ((sip >> 24) & 0xFF) | 0;
                    let dataPtr = ((sip & 0x00FFFFFF)) | 0;
                    let city = _getLong(this.totalInMemoryBytes, dataPtr);

                    const bufArray = new Array();
                    for (let startPos = dataPtr + 4, i = startPos; i < startPos + dataLen - 4; ++i) {
                        bufArray.push(this.totalInMemoryBytes[i]);
                    }
                    const region = Buffer.from(bufArray, 0).toString();
                    return { city, region };
                }
            }
        }
    }

    /**
     * Async of MemorySearch.
     * @param {String} ip 
     */
    memorySearch(ip, callBack) {

        let _ip = _ip2long(ip);
        let l = 0, h = this.totalBlocks;
        let sip = 0;
        let m = 0, p = 0;
        let self = this;

        function _innerMemorySearchLoop() {

            if (l <= h) {

                m = (l + h) >> 1;
                p = (self.firstIndexPtr + m * INDEX_BLOCK_LENGTH) | 0;

                sip = _getLong(self.totalInMemoryBytes, p);

                if (_ip < sip) {
                    h = m - 1;
                    setImmediate(_innerMemorySearchLoop);
                }
                else {
                    sip = _getLong(self.totalInMemoryBytes, p + 4);

                    if (_ip > sip) {
                        l = m + 1;
                        setImmediate(_innerMemorySearchLoop);
                    }
                    else {
                        sip = _getLong(self.totalInMemoryBytes, p + 8);
                        //not matched
                        if (sip === 0) return callBack(null, null);

                        //get the data
                        let dataLen = ((sip >> 24) & 0xFF) | 0;
                        let dataPtr = ((sip & 0x00FFFFFF)) | 0;
                        let city = _getLong(self.totalInMemoryBytes, dataPtr);

                        const bufArray = new Array();
                        for (let startPos = dataPtr + 4, i = startPos; i < startPos + dataLen - 4; ++i) {
                            bufArray.push(self.totalInMemoryBytes[i]);
                        }
                        const region = Buffer.from(bufArray).toString();
                        callBack(null, { city, region });
                    }
                }
            }
            else {
                callBack(null, null);
            }
        }

        if (this.totalInMemoryBytes === null) {

            this.totalInMemoryBytes = Buffer.alloc(this.totalInMemoryBytesSize);

            fs.read(this.dbFd, this.totalInMemoryBytes, 0, this.totalInMemoryBytesSize, 0, (err) => {
                if (err) {
                    callBack(err, null);
                }
                else {
                    this.firstIndexPtr = _getLong(this.totalInMemoryBytes, 0);
                    this.lastIndexPtr = _getLong(this.totalInMemoryBytes, 4);
                    this.totalBlocks = ((this.lastIndexPtr - this.firstIndexPtr) / INDEX_BLOCK_LENGTH) | 0 + 1;

                    _innerMemorySearchLoop();
                }
            });
        }
        else {
            _innerMemorySearchLoop();
        }
    }
}
//#endregion

module.exports = IP2Region;

File: v1.0/binding/lua/Ip2region.lua
--[[
Ip2region lua binding

@author chenxin<chenxin619315@gmail.com>
@date   2018/10/02
]]--

require("bit32");

local INDEX_BLOCK_LENGTH  = 12;
local TOTAL_HEADER_LENGTH = 8192;
local _M = {
    dbFile = "",
    dbFileHandler = "",
    dbBinStr = "",
    HeaderSip = "", 
    HeaderPtr = "",
    headerLen = 0,
    firstIndexPtr = 0,
    lastIndexPtr = 0, 
    totalBlocks = 0
};

_G["Ip2region"] = _M;

-- set the __index to itself
_M.__index = _M;

-- set the print meta-method
_M.__tostring = function(table)
    local t = {
        "dbFile=" .. table.dbFile,
        -- "dbFileHandler=" .. table.dbFileHandler,
        "headerLen=" .. table.headerLen,
        "firstIndexPtr" .. table.firstIndexPtr,
        "lastIndexPtr" .. table.lastIndexPtr,
        "totalBlocks" .. table.totalBlocks
    };

    return table.concat(t, ",");
end

--[[
construct method

@param  obj
@return Ip2region object
--]]
function _M.new(dbFile)
    obj = {};
    setmetatable(obj, _M);
    obj.dbFile = dbFile;
    return obj;
end


--[[
internal function to get a integer from a binary string

@param  dbBinStr
@param  idx
@return Integer
]]--
function getLong(bs, idx)
    local a1 = string.byte(string.sub(bs, idx, idx));
    local a2 = bit32.lshift(string.byte(string.sub(bs, idx+1, idx+1)),  8);
    local a3 = bit32.lshift(string.byte(string.sub(bs, idx+2, idx+2)), 16);
    local a4 = bit32.lshift(string.byte(string.sub(bs, idx+3, idx+3)), 24);

    local val = bit32.bor(a1, a2);
    val = bit32.bor(val, a3);
    val = bit32.bor(val, a4);

    return val;
end


--[[
internal function to convert the string ip to a long value

@param  ip
@return Integer
]]--
function _M.ip2long(self, ip)
    -- dynamic arguments checking
    -- to support object.ip2long and object:ip2long access
    if ( type(self) == "string") then
        ip = self;
    end

    local ini = 1;
    local iip = 0;
    local off = 24;
    while true do
        local pos = string.find(ip, '.', ini, true);
        if ( not pos ) then
            break;
        end

        local sub = string.sub(ip, ini, pos - 1);
        if ( string.len(sub) < 1 ) then
            return nil;
        end

        iip = bit32.bor(iip, bit32.lshift(tonumber(sub), off));
        ini = pos + 1;
        off = off - 8;
    end

    -- check if it is a valid ip address
    if ( off ~= 0 or ini > string.len(ip) ) then
        return nil;
    end

    local sub = string.sub(ip, ini);
    if ( string.len(sub) < 1 ) then
        return nil;
    end

    return bit32.bor(iip, bit32.lshift(tonumber(sub), off));
end


--[[
internal function to get the whole content of a file

@param  file
@return String
]]--
function get_file_contents(file)
    local fi = io.input(file);
    if ( not fi ) then
        return nil;
    end

    local str = io.read("*a");
    io.close();
    return str;
end


--[[
set the current db file path

@param  dbFile
]]--
function _M:setDbFile(dbFile)
    self.dbFile = dbFile;
end


--[[
all the db binary string will be loaded into memory
then search the memory only and this will a lot faster than disk base search
@Note: invoke it once before put it to public invoke could make it thread safe

@param  ip
@return table or nil for failed
]]--
function _M:memorySearch(ip)
    -- string ip conversion
    if ( type(ip) == "string" ) then
        ip = self:ip2long(ip);
        if ( ip == nil ) then
            return nil;
        end
    end;

    -- check and load the binary string for the first time
    if ( self.dbBinStr == "" ) then
        self.dbBinStr = get_file_contents(self.dbFile);
        if ( not self.dbBinStr ) then
            return nil;
        end

        self.firstIndexPtr = getLong(self.dbBinStr, 1);
        self.lastIndexPtr  = getLong(self.dbBinStr, 5);
        self.totalBlocks   = (self.lastIndexPtr - self.firstIndexPtr)/INDEX_BLOCK_LENGTH + 1;
    end


    -- binary search to define the data
    local l = 0;
    local h = self.totalBlocks;
    local dataPtr = 0;
    while ( l <= h ) do
        local m = math.floor((l + h) / 2);
        local p = self.firstIndexPtr + m * INDEX_BLOCK_LENGTH;
        local sip = getLong(self.dbBinStr, p + 1);
        if ( ip < sip ) then
            h = m - 1;
        else
            local eip = getLong(self.dbBinStr, p + 5);  -- 4 + 1
            if ( ip > eip ) then
                l = m + 1;
            else
                dataPtr = getLong(self.dbBinStr, p + 9); -- 8 + 1
                break;
            end
        end
    end

    -- not matched just stop it here
    if ( dataPtr == 0 ) then return nil end

    -- get the data
    local dataLen = bit32.band(bit32.rshift(dataPtr, 24), 0xFF);
    dataPtr = bit32.band(dataPtr, 0x00FFFFFF);
    local dptr = dataPtr + 5;   -- 4 + 1

    return {
        city_id = getLong(self.dbBinStr, dataPtr), 
        region  = string.sub(self.dbBinStr, dptr, dptr + dataLen - 5)
    };
end


--[[
get the data block through the specified ip address 
or long ip numeric with binary search algorithm

@param  ip
@return table or nil for failed
]]--
function _M:binarySearch(ip)
    -- check and conver the ip address
    if ( type(ip) == "string" ) then
        ip = self:ip2long(ip);
        if ( ip == nil ) then
            return nil;
        end
    end

    if ( self.totalBlocks == 0 ) then
        -- check and open the original db file
        self.dbFileHandler = io.open(self.dbFile, "r");
        if ( not self.dbFileHandler ) then
            return nil;
        end

        self.dbFileHandler:seek("set", 0);
        local superBlock = self.dbFileHandler:read(8);

        self.firstIndexPtr = getLong(superBlock, 1);    -- 0 + 1
        self.lastIndexPtr  = getLong(superBlock, 5);    -- 4 + 1
        self.totalBlocks   = (self.lastIndexPtr-self.firstIndexPtr)/INDEX_BLOCK_LENGTH + 1;
    end


    -- binary search to define the data
    local l = 0;
    local h = self.totalBlocks;
    local dataPtr = 0;
    while ( l <= h ) do
        local m = math.floor((l + h) / 2);
        local p = m * INDEX_BLOCK_LENGTH;
        self.dbFileHandler:seek("set", self.firstIndexPtr + p);
        local buffer = self.dbFileHandler:read(INDEX_BLOCK_LENGTH);
        local sip = getLong(buffer, 1); -- 0 + 1
        if ( ip < sip ) then
            h = m - 1;
        else
            local eip = getLong(buffer, 5);   -- 4 + 1
            if ( ip > eip ) then
                l = m + 1;
            else
                dataPtr = getLong(buffer, 9);   -- 8 + 1
                break;
            end
        end
    end

    -- not matched just stop it here
    if ( dataPtr == 0 ) then return nil; end


    -- get the data
    local dataLen = bit32.band(bit32.rshift(dataPtr, 24), 0xFF);
    dataPtr = bit32.band(dataPtr, 0x00FFFFFF);

    self.dbFileHandler:seek("set", dataPtr);
    local data = self.dbFileHandler:read(dataLen);

    return {
        city_id = getLong(data, 1),    -- 0 + 1
        region  = string.sub(data, 5)  -- 4 + 1
    };
end


--[[
get the data block associated with the specified ip with b-tree search algorithm

@param  ip
@return table or nil for failed
]]--
function _M:btreeSearch(ip)
    -- string ip to integer conversion
    if ( type(ip) == "string" ) then
        ip = self:ip2long(ip);
        if ( ip == nil ) then
            return nil;
        end
    end

    -- check and load the header
    if ( self.headerLen == 0 ) then
        -- check and open the original db file
        self.dbFileHandler = io.open(self.dbFile, 'r');
        if ( not self.dbFileHandler ) then
            return nil;
        end

        self.dbFileHandler:seek("set", 8);
        local buffer = self.dbFileHandler:read(TOTAL_HEADER_LENGTH);
        
        -- fill the header
        local i = 0;
        local idx = 0;
        self.HeaderSip = {};
        self.HeaderPtr = {};
        for i=0, TOTAL_HEADER_LENGTH, 8 do
            local startIp = getLong(buffer, i + 1); -- 0 + 1
            local dataPtr = getLong(buffer, i + 5); -- 4 + 1
            if ( dataPtr == 0 ) then
                break;
            end

            table.insert(self.HeaderSip, startIp);
            table.insert(self.HeaderPtr, dataPtr);
            idx = idx + 1;
        end

        self.headerLen = idx;
    end
    

    -- 1. define the index block with the binary search
    local l = 0; 
    local h = self.headerLen;
    local sptr = 0;
    local eptr = 0;
    while ( l <= h ) do
        local m = math.floor((l + h) / 2);
        -- perfetc matched, just return it
        if ( ip == self.HeaderSip[m] ) then
            if ( m > 0 ) then
                sptr = self.HeaderPtr[m-1];
                eptr = self.HeaderPtr[m  ];
            else
                sptr = self.HeaderPtr[m  ];
                eptr = self.HeaderPtr[m+1];
            end
            
            break;
        end
        
        -- less then the middle value
        if ( ip < self.HeaderSip[m] ) then
            if ( m == 0 ) then
                sptr = self.HeaderPtr[m  ];
                eptr = self.HeaderPtr[m+1];
                break;
            elseif ( ip > self.HeaderSip[m-1] ) then
                sptr = self.HeaderPtr[m-1];
                eptr = self.HeaderPtr[m  ];
                break;
            end
            h = m - 1;
        else
            if ( m == self.headerLen - 1 ) then
                sptr = self.HeaderPtr[m-1];
                eptr = self.HeaderPtr[m  ];
                break;
            elseif ( ip <= self.HeaderSip[m+1] ) then
                sptr = self.HeaderPtr[m  ];
                eptr = self.HeaderPtr[m+1];
                break;
            end
            l = m + 1;
        end
    end
    
    -- match nothing just stop it
    if ( sptr == 0 ) then return nil; end
    
    -- 2. search the index blocks to define the data
    self.dbFileHandler:seek("set", sptr);
    local blockLen = eptr - sptr;
    local index = self.dbFileHandler:read(blockLen + INDEX_BLOCK_LENGTH);
    local dataPtr = 0;

    l = 0;
    h = blockLen / INDEX_BLOCK_LENGTH;
    while ( l <= h ) do
        local m = math.floor((l + h) / 2);
        local p = m * INDEX_BLOCK_LENGTH;
        local sip = getLong(index, p + 1);       -- 0 + 1
        if ( ip < sip ) then
            h = m - 1;
        else
            local eip = getLong(index, p + 5);   -- 4 + 1
            if ( ip > eip ) then
                l = m + 1;
            else
                dataPtr = getLong(index, p + 9); -- 8 + 1
                break;
            end
        end
    end
    
    -- not matched
    if ( dataPtr == 0 ) then return nil; end
    
    -- 3. get the data
    local dataLen = bit32.band(bit32.rshift(dataPtr, 24), 0xFF);
    dataPtr = bit32.band(dataPtr, 0x00FFFFFF);
    
    self.dbFileHandler:seek("set", dataPtr);
    local data = self.dbFileHandler:read(dataLen);

    return {
        city_id = getLong(data, 1),     -- 0 + 1
        region  = string.sub(data, 5)   -- 4 + 1
    };
end


--[[
close the object and do the basic gc
]]--
function _M.close(self)
    if ( self.dbFileHandler ~= "" ) then
        self.dbFileHandler:close();
    end

    if ( self.dbBinStr ~= "" ) then
        self.dbBinStr = nil;
    end
end


return _M;
